Agent skills for Jaspr (#778)
schultek/jaspr: packages/jaspr/skills/jaspr-styling/SKILL.md

---
name: jaspr-styling

> description: Guide to writing native structural Jaspr CSS. Use this skill when styling components, implementing themes, or using CSS properties.

Reviewer: Seems it's best to use imperative language or "Use when/if".

```suggestion
description: Write type-safe CSS-in-Dart to style Jaspr components. Use this skill when styling components, implementing themes, or using CSS properties.
```

  }

  @css
  static List<StyleRule> get styles => [
    css('.main', [
      // Use css('&') to refer to the parent selector (.main)
      css('&').styles(
        width: 100.px, // Note: Use `.px`, `.rem`, `.percent` extensions on numbers.
        padding: .all(10.rem),
        display: .flex,
      ),
      css('&:hover').styles(
        backgroundColor: Colors.blue,
      ),
      css('p').styles(
        color: Colors.blue,
      ),
    ]),
    // Responsive queries
    css.media(MediaQuery.screen(maxWidth: 600.px), [
      css('.main').styles(flexDirection: .column),
    ])
  ];
}
```

### 2. Inline Styles

You can pass `Styles` instances directly to native HTML components. This renders as the `style="..."` HTML attribute.

- **Rule 1:** You MUST ONLY use inline styles for **dynamic styles** (e.g., styles that change based on state or parameters).
- **Rule 2:** For static styles, you MUST use the component-level `@css` annotation.
- **Rule 3:** You MUST NOT use inline styles for complex rules like media queries, hover states, or animations.

```dart
// Example of dynamic inline styling driven by state
class ColorBox extends StatelessComponent {
  final Color boxColor;

>   const ColorBox({required this boxColor, super.key});

Reviewer: ```suggestion
  const ColorBox({required this.boxColor, super.key});
```

  @override
  Component build(BuildContext context) {
    return div(styles: Styles(backgroundColor: boxColor), []);
  }
}
```

### 3. Global `@css` Styles

You can define a global set of styles directly from Dart using the `@css` annotation on a global variable or getter.

- **Rule 1:** Just like component-level `@css`, global `@css` is ONLY supported in **server** and **static** modes.
- **Rule 2:** You MUST place global `@css` annotations on top-level getters or static getters returning `List<StyleRule>`.

```dart
@css
List<StyleRule> get globalStyles => [
  css('body').styles(
    margin: .zero,
    fontFamily: .list([FontFamily('Roboto'), FontFamilies.sansSerif]),
    backgroundColor: Colors.white,
  ),
  css('a').styles(
    textDecoration: TextDecoration(line: .none),
    color: Colors.blue,
  ),
];
```

### 4. Global External Stylesheets

- **Rule:** If you are using raw `.css` files (or other css frameworks like sass/scss, tailwind, etc.), you MUST include them using a `<link rel="stylesheet">` element inside the `Document(head: [...])` component (for server/static mode) or in `web/index.html` (for client mode).

---

> ## 🛠 Resource Strategy: Jaspr Styles Properties

Reviewer: Can simplify this:

```suggestion
## Jaspr Styles Properties
```

  TextAlign? textAlign,
  FontFamily? fontFamily,
  Unit? fontSize,
  FontWeight? fontWeight,
  FontStyle? fontStyle,
  TextDecoration? textDecoration,
  TextTransform? textTransform,
  Unit? textIndent,
  Unit? letterSpacing,
  Unit? wordSpacing,
  Unit? lineHeight,
  TextShadow? textShadow,
  TextOverflow? textOverflow,
  WhiteSpace? whiteSpace,
  Quotes? quotes,
  // Background Styles
  Color? backgroundColor,
  ImageStyle? backgroundImage,
  BackgroundOrigin? backgroundOrigin,
  BackgroundPosition? backgroundPosition,
  BackgroundAttachment? backgroundAttachment,
  BackgroundRepeat? backgroundRepeat,
  BackgroundSize? backgroundSize,
  BackgroundClip? backgroundClip,
  // Raw Styles
  Map<String, String>? raw,
})
```

- **Rule 1:** You MUST define all properties in the order they are defined in the `Styles` class.
- **Rule 2:** You MUST use dot-shorthands for all style properties and values where applicable (e.g., `padding: .all(10.px)` instead of `padding: Padding.all(Unit.pixels(10))`, or `justifyContent: .center` instead of `justifyContent: JustifyContent.center`).
- **Rule 3:** You MUST use `raw` for any properties that are not supported by the `Styles` class.

> **IMPORTANT:** When you are unsure about a specific layout mechanism, you **MUST** read the respective reference file provided alongside this skill:

Reviewer: Seems the following files cover more than just layout.

```suggestion
**IMPORTANT:** When you are unsure about a specific styling mechanism, you **MUST** read the respective reference file provided alongside this skill:
```
